internal/jobcontainers/job.go

```diff
@@ -0,0 +1,364 @@
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
+}
+
+// Terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *jobObject) terminate() error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+	return windows.TerminateJobObject(job.jobHandle, 1)
+}
+
+func (job *jobObject) shutdown(ctx context.Context) error {
+	if job.jobHandle == 0 {
+		return errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	pids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	var (
+		terminate bool
+		signalErr bool
+	)
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Error("failed to send ctrl-break to process in job")
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again. If there is any left then terminate the job.
+	newPids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	terminate = len(newPids) != 0
+	// If any of the processes couldnt be killed gracefully just terminate the job.
+	// Equivalent to calling TerminateProcess on every proc in the job.
+	if terminate || signalErr {
+		return job.terminate()
+	}
+	return nil
+}
+
+// Returns all of the process IDs in the job object.
+func (job *jobObject) pids() ([]uint32, error) {
+	if job.jobHandle == 0 {
+		return nil, errAlreadyClosed
+	}
+	job.handleLock.RLock()
+	defer job.handleLock.RUnlock()
+
+	info := winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST{}
+	err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		nil,
+	)
+
+	// If there are no processes in the job the next call to QueryInformation will just
+	// hang until a memory alloc error. Return here.
+	if err == nil {
+		// Return empty slice instead of nil to play well with the caller of this.
+		// Do not return an error if no processes are running inside the job
+		return []uint32{}, nil
+	}
+
+	if err != winapi.ERROR_MORE_DATA {
+		return nil, fmt.Errorf("failed initial query for PIDs in job object: %s", err)
+	}
+
+	jobBasicProcessIDListSize := unsafe.Sizeof(info) + (unsafe.Sizeof(info.ProcessIdList[0]) * uintptr(info.NumberOfAssignedProcesses-1))
+	buf := make([]byte, jobBasicProcessIDListSize)
+	if err = winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicProcessIdList,
+		uintptr(unsafe.Pointer(&buf[0])),
+		uint32(len(buf)),
+		nil,
+	); err != nil {
+		return nil, fmt.Errorf("failed to query for PIDs in job object: %s", err)
+	}
+
+	bufInfo := (*winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST)(unsafe.Pointer(&buf[0]))
+	bufPids := make([]uintptr, bufInfo.NumberOfProcessIdsInList)
+	if err = winapi.RtlMoveMemory(
```

Comment:
I think we can avoid this allocation, and just walk the list and directly assign to the final `[]uint32` slice.

Reply:
Can you share how you'd do this with how we have the definitions currently? ProcessIdList is just an array of size 1, we could expand this to some arbitrarily large value like 100, 1000 etc. and just walk up to bufinfo.NumberOfProcessIdsInList or is there some cast that we can do thats eluding me?

Reply:
We should just be able to cast this into a pointer to a `uintptr` array. It also might be nice to provide a `AllProcessIds` method on `winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST`, that returns the slice. Here's an example of how this is done for another type in `golang.org/x/sys/windows`: https://github.com/golang/sys/blob/d9f96fdee20d/windows/security_windows.go#L613

Reply:
Oh sweet, I'll change to this then. Thanks Kevin